README.md
@gavant/ember-button-basic
==============================================================================

A simple and flexible `<Button>` Glimmer component, with built-in Bootstrap 4 and FontAwesome 5 integration.


Compatibility
------------------------------------------------------------------------------

* Ember.js v3.24 or above
* Ember CLI v3.24 or above
* Node.js v12 or above


Installation
------------------------------------------------------------------------------

```
ember install @gavant/ember-button-basic
```

**NOTE:** This addon uses FontAwesome 5 ([@fortawesome/ember-fontawesome](https://github.com/FortAwesome/ember-fontawesome)) for icon support. However it does NOT install any icon set packages. You must install these separately, following the addon's installation guide, e.g.
```
yarn add --dev @fortawesome/free-solid-svg-icons
```


Usage
------------------------------------------------------------------------------

An example `<Button>` usage, with all available arguments used. None of the component arguments are required, but in most cases you'll either want to provide a `@label` and/or `@icon`, or block content, and a click `@action`.

```hbs
<Button
    @action={{function}}
    @label={{string}}
    @buttonType={{'button' | 'submit' | 'reset'}}
    @type={{'primary' | 'secondary' | 'success' | 'info' | 'warning' | 'danger' | 'link' | 'light' | 'dark'}}
    @size={{'lg' | 'md' | 'sm' | 'xs'}}
    @blockLayout={{true | false}}
    @wide={{true | false}}
    @outline={{true | false}}
    @rounded={{true | false}}
    @circle={{true | false}}
    @iconOnly={{true | false}}
    @active={{true | false}}
    @disabled={{true | false}}
    @preventDefault={{true | false}}
    @stopPropagation={{true | false}}
    @icon={{string}}
    @iconPrefix={{string}}
    @iconClass={{string}}
    @iconSize={{'xs' | 'sm' | 'lg' | '2x' | '3x' | '4x' | '5x' | '6x'|  '7x' | '8x' | '9x' | '10x'}}
    @iconFlip={{'horizontal' | 'vertical' | 'both'}}
    @iconRotation={{90 | 180 | 270}}
    @iconSpin={{true | false}}
    @iconBorder={{true | false}}
    @iconFixedWidth={{true | false}}
/>

{{!-- block form is supported too --}}
<Button @action={{this.someAction}}>
    Look ma, block content!
</Button>
```

Using standard Glimmer component conventions, you can easily extend this component to add more advanced functionality. For example:

```hbs
{{!-- app/components/button-with-tooltip --}}
<Button 
    @action={{@action}}
    @label={{@label}} 
    @type={{@type}}
    ...attributes
>
    {{yield}}
    {{#if @tooltip}}
        <EmberTooltip @side="right">
            {{@tooltip}}
        </EmberTooltip>
    {{/if}}
</Button>


{{!-- elsewhere in your app.., --}}
<ButtonWithTooltip @label="I have a toolip" @tooltip="This is the tooltip!" />
```


Contributing
------------------------------------------------------------------------------

See the [Contributing](CONTRIBUTING.md) guide for details.


License
------------------------------------------------------------------------------

This project is licensed under the [MIT License](LICENSE.md).


CHANGELOG.md




## v3.0.4 (2022-07-26)

## v3.0.3 (2022-07-26)

#### :bug: Bug Fix
* [#36](https://github.com/Gavant/gavant-ember-button-basic/pull/36) Fixing node16 types ([@bakerac4](https://github.com/bakerac4))

#### Committers: 1
- Adam Baker ([@bakerac4](https://github.com/bakerac4))

## v3.0.2 (2022-07-26)

#### :bug: Bug Fix
* [#35](https://github.com/Gavant/gavant-ember-button-basic/pull/35) Fixing glint ([@bakerac4](https://github.com/bakerac4))

#### Committers: 1
- Adam Baker ([@bakerac4](https://github.com/bakerac4))

## v3.0.1 (2022-07-26)

#### :bug: Bug Fix
* [#34](https://github.com/Gavant/gavant-ember-button-basic/pull/34) Fix types ([@bakerac4](https://github.com/bakerac4))

#### Committers: 1
- Adam Baker ([@bakerac4](https://github.com/bakerac4))

## v3.0.0 (2022-07-22)

#### :boom: Breaking Change
* [#33](https://github.com/Gavant/gavant-ember-button-basic/pull/33) Converting addon to v2 ([@bakerac4](https://github.com/bakerac4))

#### Committers: 1
- Adam Baker ([@bakerac4](https://github.com/bakerac4))

#### [v2.0.0](https://github.com/Gavant/gavant-ember-button-basic/compare/v1.0.1...v2.0.0)

-   Upgrade to 4.0 [`#31`](https://github.com/Gavant/gavant-ember-button-basic/pull/31)
-   Starting upgrade process [`dd983fc`](https://github.com/Gavant/gavant-ember-button-basic/commit/dd983fc625587d8e082446edbee0b98a38e34670)
-   v3.16.0...v4.3.0 [`7abc52a`](https://github.com/Gavant/gavant-ember-button-basic/commit/7abc52aa4e702a2bd6b04b5354d15d0bc39deeed)
-   Pin node [`942fec0`](https://github.com/Gavant/gavant-ember-button-basic/commit/942fec0944be9a5bb743c13c6d6cf1e2d1fb7100)

#### [v1.0.1](https://github.com/Gavant/gavant-ember-button-basic/compare/v1.0.0...v1.0.1)

> 30 March 2020

-   all generic args in the Button component [`ccf04a7`](https://github.com/Gavant/gavant-ember-button-basic/commit/ccf04a7d52832207693efb723d305f7c35157a66)

#### v1.0.0

> 23 March 2020

-   Octanification: TypeScript Strikes Back [`#6`](https://github.com/Gavant/gavant-ember-button-basic/pull/6)
-   Avoid 3rd party helper naming collision [`#3`](https://github.com/Gavant/gavant-ember-button-basic/pull/3)
-   Additional attribute bindings [`#2`](https://github.com/Gavant/gavant-ember-button-basic/pull/2)
-   Misc cleanup [`#1`](https://github.com/Gavant/gavant-ember-button-basic/pull/1)
-   Initial set up of Gavants button-basic [`2808712`](https://github.com/Gavant/gavant-ember-button-basic/commit/28087121a258abc41b0b54519e9a3274f1102a11)
-   Initial Commit from Ember CLI v3.1.3 [`e93401b`](https://github.com/Gavant/gavant-ember-button-basic/commit/e93401b755032d8587d8126f0507d5ae8cf80999)
-   update yarn.lock [`7d1bb0a`](https://github.com/Gavant/gavant-ember-button-basic/commit/7d1bb0a93fc64b67cf3e4beaeadfd35ff4f75763)


addon/components/button.ts
import Component from '@glimmer/component';
import { action } from '@ember/object';

export interface ButtonArgs {
    //this needs to be an arg and not just passed in as an HTML element like `type="submit"` due to an open bug
    //@see https://github.com/emberjs/ember.js/issues/18232
    buttonType?: 'button' | 'submit' | 'reset';
    type?: 'primary' | 'secondary' | 'success' | 'info' | 'warning' | 'danger' | 'link' | 'light' | 'dark';
    size?: 'lg' | 'md' | 'sm' | 'xs';
    blockLayout?: boolean;
    wide?: boolean;
    outline?: boolean;
    rounded?: boolean;
    circle?: boolean;
    iconOnly?: boolean;
    active?: boolean;
    disabled?: boolean;
    label?: string;
    icon?: string;
    iconPrefix?: string;
    iconClass?: string;
    iconSize?: 'xs' | 'sm' | 'lg' | '2x' | '3x' | '4x' | '5x' | '6x'|  '7x' | '8x' | '9x' | '10x';
    iconFlip?: 'horizontal' | 'vertical' | 'both';
    iconRotation?: 90 | 180 | 270;
    iconSpin?: boolean;
    iconBorder?: boolean;
    iconFixedWidth?: boolean;
    preventDefault?: boolean;
    stopPropagation?: boolean;
    action?: (event: Event) => void;
}

export default class Button<T extends ButtonArgs> extends Component<T> {
    /**
     * The button's HTML `type` attribute value
     */
    get buttonType() {
        return this.args.buttonType || 'button';
    }

    /**
     * Bootstrap button style type (e.g. primary, info, danger)
     */
    get type() {
        return this.args.type || 'secondary';
    }

    /**
     * Bootstrap button style class (e.g. btn-primary, btn-outline-success)
     */
    get buttonTypeClass() {
        let btnClass = '';
        if(this.args.outline) {
            btnClass = `btn-outline-${this.type}`;
        } else {
            btnClass = `btn-${this.type}`;
        }

        return btnClass;
    }

    /**
     * Bootstrap size class (e.g. btn-lg, btn-sm)
     */
    get buttonSizeClass() {
        return this.args.size ? `btn-${this.args.size}` : '';
    }

    /**
     * Buttons are given a custom class if they only contain an icon
     */
    get isIconButton() {
        return this.args.iconOnly && this.args.icon;
    }

    /**
     * Handles button click events
     * @param {Event} event
     */
    @action
    onClick(event: Event) {
        if(this.args.preventDefault) {
            event.preventDefault();
        }

        if(this.args.stopPropagation) {
            event.stopPropagation();
        }

        if(this.args.action) {
            this.args.action(event);
        }
    }
}
